### PR TITLE
feat: support more keywords and improve conversion results for complex json schema

### LINKED ISSUE #200: Bug: Error Json schema import output for Github workflow.
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)

The input json schema config: https://github.com/kcl-lang/modules/blob/main/github-workflow/jsonschema/github-workflow.json

### 2. What did you expect to see? (Required)

The right union type annotation instead of wrong type annotation. Besides, this issue can be aligned with the implementation in the kcl-lang/kcl-openapi repo.

image

### 3. What did you see instead (Required)

The wrong type annotations

image

image

image

It should be noted that the errors encountered may go far beyond those mentioned in the issue, and the input and output results should be carefully checked.

### 4. What is your KCL components version? (Required)

kcl-go v0.7.1

### PR DESCRIPTION
**Pull request by jakezhu9**
#### 1. Does this PR affect any open issues?(Y/N) and add issue references (e.g. "fix #123", "re #123".):

- [ ] N
- [x] Y 

fix #200 

#### 2. What is the scope of this PR (e.g. component or file name):

pkg/tools/gen

#### 3. Provide a description of the PR(e.g. more details, effects, motivations or doc link):

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [x] Other

Support oneOf, allOf and patternProperties keywords, improve conversion results for $ref and so on. Now we shuold get a better result when importing from complex json schema. 

I try use it to convert the josn schema of github workflow (`pkg/tools/gen/testdata/jsonschema/complex-workflow/expect.k`), and use the result to check a workflow (`pkg/tools/gen/testdata/yaml/workflow/expect.k`). After fixing some attribute name with `-` (`working-directory` -> `working_directory`, `runs-on`->`runs_on`), everything works fine!

image

Of course, the processing of some keywords (such as allOf) is not perfect yet, and we can improve it in the future.

#### 4. Are there any breaking changes?(Y/N) and describe the breaking changes(e.g. more details, motivations or doc link):

- [x] N
- [ ] Y 

#### 5. Are there test cases for these changes?(Y/N) select and add more details, references or doc links:

- [x] Unit test
- [ ] Integration test
- [ ] Benchmark (add benchmark stats below)
- [ ] Manual test (add detailed scripts or steps below)
- [ ] Other

### CODE REVIEW

File: pkg/tools/gen/testdata/jsonschema/complex-workflow/expect.k [hunk 1]
```diff
@@ -0,0 +1,1747 @@
...
+    Runs your workflow anytime the branch_protection_rule event occurs. More than one activity type triggers this event.
+
+    Attributes
+    ----------
+    types : ["created" | "edited" | "deleted"], optional, default is ["created" "edited" "deleted"]
+        Selects the types of activity that will trigger a workflow run. Most GitHub events are triggered by more than one type of activity. For example, the event for the release resource is triggered when a release is published, unpublished, created, edited, deleted, or prereleased. The types keyword enables you to narrow down activity that causes the workflow to run. When only one activity type triggers a webhook event, the types keyword is unnecessary.
+        You can use an array of event types. For more information about each event and their activity types, see https://help.github.com/en/articles/events-that-trigger-workflows#webhook-events.
+    """
+
+    types?: ["created" | "edited" | "deleted"] = ["created" "edited" "deleted"]
```
Comment (Peefy) [inline pkg/tools/gen/testdata/jsonschema/complex-workflow/expect.k]: The default value `["created" "edited" "deleted"]` may have better format ["created", "edited", "deleted"].

File: pkg/tools/gen/template.go [hunk 2]
```diff
@@ -103,6 +103,15 @@ func formatValue(v interface{}) string {
 			s.WriteString(fmt.Sprintf("%s: %s", formatValue(key), formatValue(value[key])))
 		}
 		return "{" + s.String() + "}"
+	case []interface{}:
+		var s strings.Builder
+		for i, item := range value {
+			if i != 0 {
+				s.WriteString(" ")
```
Comment (Peefy) [inline pkg/tools/gen/template.go]: s.WriteString(",")

File: pkg/tools/gen/testdata/jsonschema/complex-workflow/expect.k (hunk 1 above)
Comment (jakezhu9) [reply pkg/tools/gen/testdata/jsonschema/complex-workflow/expect.k]: Done

File: pkg/tools/gen/testdata/jsonschema/unsupport-multi-pattern-props/expect.k [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+    authors?: BookAuthors
+
+schema BookAuthors:
+    r"""
+    BookAuthors
+    """
+
+    [...str]: any
+
+schema BookAuthorsPatternProperties:
```
Comment (Peefy) [inline pkg/tools/gen/testdata/jsonschema/unsupport-multi-pattern-props/expect.k]: Why generate only one schema here? I see two pattern objects in the json schema.

Comment (jakezhu9) [reply pkg/tools/gen/testdata/jsonschema/unsupport-multi-pattern-props/expect.k]: You are right, we should generate PatternProperties0, PatternProperties1...

File: pkg/tools/gen/testdata/jsonschema/unsupport-multi-pattern-props/expect.k [hunk 4]
```diff
@@ -0,0 +1,49 @@
+"""
```
Comment (Peefy) [inline pkg/tools/gen/testdata/jsonschema/unsupport-multi-pattern-props/expect.k:1]: TestGenKclFromJsonSchema/unsupport-multi-pattern-props test failed in CI on windows.

Comment (jakezhu9) [reply pkg/tools/gen/testdata/jsonschema/unsupport-multi-pattern-props/expect.k:1]: Fixed

**Comment by Peefy** [comment]
Good Job! LGTM!

**Review by Peefy — APPROVED** [review]
LGTM

**Pull request merged**